framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java

```diff
@@ -957,6 +960,73 @@ public Void visitMethodInvocation(MethodInvocationTree node, Void p) {
     }
 
     /**
+     * A helper method to check that the array type of actual varargs is a subtype of the
+     * corresponding required varargs, and issues "argument.invalid" error if it's not a subtype of
+     * the required one.
+     *
+     * <p>Note it's required that type checking for the each elements of varargs is executed by
+     * caller before or after calling this method.
+     *
+     * @see #checkArguments(List, List)
+     * @param invokedMethod the method type to be invoked
+     * @param tree method or constructor invocation tree
+     */
+    private void checkVarargs(AnnotatedExecutableType invokedMethod, Tree tree) {
+        if (!invokedMethod.isVarArgs()) {
+            return;
+        }
+
+        List<AnnotatedTypeMirror> formals = invokedMethod.getParameterTypes();
+        int numFormals = formals.size();
+        int lastArgIndex = numFormals - 1;
+        AnnotatedArrayType lastParamAnnotatedType = (AnnotatedArrayType) formals.get(lastArgIndex);
+
+        // We will skip type checking so that we avoid duplicating error message
+        // if the last argument is same depth with the depth of formal varargs
+        // because type checking is already done in checkArguments.
+        List<? extends ExpressionTree> args;
+        switch (tree.getKind()) {
+            case METHOD_INVOCATION:
+                args = ((MethodInvocationTree) tree).getArguments();
+                break;
+            case NEW_CLASS:
+                args = ((NewClassTree) tree).getArguments();
+                break;
+            default:
+                throw new AssertionError("Unexpected kind of tree: " + tree);
+        }
+        if (numFormals == args.size()) {
+            AnnotatedTypeMirror lastArgType =
+                    atypeFactory.getAnnotatedType(args.get(args.size() - 1));
+            if (lastArgType.getKind() == TypeKind.ARRAY
+                    && getArrayDepth(lastParamAnnotatedType)
+                            == getArrayDepth((AnnotatedArrayType) lastArgType)) {
+                return;
+            }
+        }
+
+        AnnotatedTypeMirror wrappedVarargsType = atypeFactory.getAnnotatedTypeVarargsArray(tree);
+
+        // When dataflow analysis is not enabled, it will be null and we can suppose to there is no
+        // annotation to be checked for generated varargs array.
+        if (wrappedVarargsType == null) {
+            return;
+        }
+
+        // If the component type is intersection or including type variable by type inference,
```

Comment:
The comment and the code seem disconnected - I don't see a check for intersection types or type variables in the following code. Can you clarify the comment?